theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,686 @@
+register_specialize_device(local_conv2d_corrmm)
+
+@local_optimizer([AbstractConv2d_gradWeights])
+def local_conv2d_gradweight_corrmm(node):
+
+    img, topgrad, shape = node.inputs
+
+    if not isinstance(img.type, CudaNdarrayType) or \
+            not isinstance(topgrad.type, CudaNdarrayType):
+        return None
+    if node.op.filter_flip:
+        img = img[:, :, ::-1, ::-1]
+    rval = GpuCorrMM_gradWeights(border_mode=node.op.border_mode,
+                                 subsample=node.op.subsample)(
+        gpu_contiguous(img), gpu_contiguous(topgrad), shape)
+    return [rval]
+register_specialize_device(local_conv2d_gradweight_corrmm)
+
+@local_optimizer([AbstractConv2d_gradInputs])
+def local_conv2d_gradinputs_corrmm(node):
+    kern, topgrad, shape = node.inputs
+
+    if not isinstance(kern.type, CudaNdarrayType) or \
+            not isinstance(topgrad.type, CudaNdarrayType):
+        return None
+
+    if node.op.filter_flip:
+        kern = kern[:, :, ::-1, ::-1]
+
+    rval =  GpuCorrMM_gradInputs(border_mode=node.op.border_mode,
+    subsample=node.op.subsample)(
+        gpu_contiguous(kern), gpu_contiguous(topgrad), shape)
+    return [rval]
+register_specialize_device(local_conv2d_gradinputs_corrmm)
+
+
+
+### Cpu Optmization
+### Desactived focus on GPU optimization first
+@local_optimizer([AbstractConv2d])
+def local_conv2d_cpu(node):
+
+    if not isinstance(node.op, AbstractConv2d):
+        return None
+
+    img, kern = node.inputs
+    if isinstance(img.type, CudaNdarrayType) or \
+            isinstance(kern.type, CudaNdarrayType):
+        return None
+
+    rval = cpu_conv2d(img, kern,
+                      node.op.imshp, node.op.kshp,
+                      border_mode=node.op.border_mode,
+                      subsample=node.op.subsample)
+    return [rval]
+register_specialize_device(local_conv2d_cpu)
+
+
+@local_optimizer([AbstractConv2d_gradWeights])
+def local_conv2d_gradweight_cpu(node):
+
+    ## len is 4 all the time
+    img, topgrad, shape = node.inputs
+    if isinstance(img.type, CudaNdarrayType) or \
+            isinstance(topgrad.type, CudaNdarrayType):
+        return None
+
+    if node.op.border_mode == 'valid' and node.op.subsample != (1, 1):
+        # Use the gradient as defined in conv3D, because the implementation
+        # by Conv is slow (about 3x slower than conv3D, and probably 10x
+        # slower than it could be), nad incorrect when dx or dy > 2.
+        # build a "node", that should be equivalent to the one given by
+        # self.make_node, but using convGrad3D instead.
+        shuffled_img = img.dimshuffle(0, 2, 3, 'x', 1)
+        shuffled_topgrad = topgrad.dimshuffle(0, 2, 3, 'x', 1)
+        rval = ConvGrad3D(V=shuffled_img,
+                          d=(op.subsample[0], op.subsample[1], 1),
+                          WShape=(self.kshp[0], self.kshp[1], 1),
+                          dCdH_=shuffled_topgrad)
+
+        return [rval.dimshuffle(0, 4, 1, 2)]
```

Comment:
If flip_kernel is True, that returned value should be flipped as well.
If flip_kernel is False, I'm not sure if something else has to be flipped somewhere.